daemon/api_download.go

```diff
@@ -102,55 +108,154 @@ func postSnapDownload(c *Command, r *http.Request, user *auth.UserState) Respons
+			Action:       "download",
+			InstanceName: action.SnapName,
+			Revision:     action.Revision,
+			CohortKey:    action.CohortKey,
+			Channel:      action.Channel,
+		}}
+		sars, err := theStore.SnapAction(context.TODO(), nil, actions, user, nil)
+		if err != nil {
+			return errToResponse(err, []string{action.SnapName}, InternalError, "cannot download snap: %v")
+		}
+		if len(sars) != 1 {
+			return InternalError("internal error: unexpected number %v of results for a single download", len(sars))
+		}
+		info = sars[0].Info
+
+		ss, err = newSnapStream(action.SnapName, info, secret)
+		if err != nil {
+			return InternalError(err.Error())
+		}
+	} else {
+		var err error
+		ss, err = newResumingSnapStream(action.SnapName, action.ResumeToken, secret)
+		if err != nil {
+			return BadRequest(err.Error())
+		}
+		ss.resume = action.resumePosition
 	}
 
-	rsp := newSnapStream(action.SnapName, info, action.resumePosition)
 	if !action.HeaderPeek {
-		r, s, err := theStore.DownloadStream(context.TODO(), action.SnapName, &downloadInfo, action.resumePosition, user)
+		r, s, err := theStore.DownloadStream(context.TODO(), action.SnapName, ss.Info, action.resumePosition, user)
 		if err != nil {
 			return InternalError(err.Error())
 		}
-		rsp.stream = r
+		ss.stream = r
 		if s != 206 {
 			// store/cdn has no partial content (valid
 			// reply per RFC)
 			logger.Debugf("store refused our range request")
-			rsp.resume = 0
+			ss.resume = 0
 		}
 	}
 
-	return rsp
+	return ss
+}
+
+func newSnapStream(snapName string, info *snap.Info, secret []byte) (*snapStream, error) {
+	dlInfo := &info.DownloadInfo
+	fname := filepath.Base(info.MountFile())
+	tokenJSON := downloadTokenJSON{
+		SnapName: snapName,
+		Filename: fname,
+		Info:     dlInfo,
+	}
+	tokStr, err := sealDownloadToken(&tokenJSON, secret)
+	if err != nil {
+		return nil, err
+	}
+	return &snapStream{
+		SnapName: snapName,
+		Filename: fname,
+		Info:     dlInfo,
+		Token:    tokStr,
+	}, nil
 }
 
-func newSnapStream(snapName string, info *snap.Info, resumePos int64) *snapStream {
+func newResumingSnapStream(snapName string, tokStr string, secret []byte) (*snapStream, error) {
+	d, err := unsealDownloadToken(tokStr, secret)
+	if err != nil {
+		return nil, err
+	}
+	if d.SnapName != snapName {
+		return nil, fmt.Errorf("resume snap name does not match original snap name")
+	}
 	return &snapStream{
 		SnapName: snapName,
-		Filename: filepath.Base(info.MountFile()),
-		Info:     info.DownloadInfo,
-		resume:   resumePos,
+		Filename: d.Filename,
+		Info:     d.Info,
+	}, nil
+}
+
+type downloadTokenJSON struct {
+	SnapName string             `json:"snap-name"`
+	Filename string             `json:"filename"`
+	Info     *snap.DownloadInfo `json:"dl-info"`
+}
+
+func sealDownloadToken(d *downloadTokenJSON, secret []byte) (string, error) {
+	b, err := json.Marshal(d)
+	if err != nil {
+		return "", err
+	}
+	mac := hmac.New(sha256.New, secret)
+	mac.Write(b)
+	tok := mac.Sum(b)
```

Comment:
Maybe a small comment here that this will result in `b+<hash>` ? I found the upstream docs not super obvious for this. I could also add this comment in a followup if this PR is green (i.e. we should not block the merge on this).

Reply:
It's green ATM, but yes something like:

// append the HMAC hash to b to build the full raw token tok

Reply:
I have the same feeling, I was not expecting `Sum` to do what it is documented to do.

In general spending a few sentences on documenting the sealing would be good.